chains/evm/contracts/libraries/USDCSourcePoolDataCodec.sol

```diff
@@ -0,0 +1,160 @@
+/// into the corresponding struct. This adds some additional gas overhead during decoding, but the benefits of saving
+/// space on the source pool data outweigh the overhead.
+/// @dev Any future versions of CCTP should include in this library a new function for encoding and decoding the source
+/// pool data accordingly.
+library USDCSourcePoolDataCodec {
+  error InvalidVersion(uint32 version);
+
+  /// @notice Encodes the source token data payload into a bytes array.
+  /// @dev By using abi.encodePacked(), significant amount of space on the source pool data is saved.
+  /// since abi.encode pads every field to the nearest 32 bytes. While it adds some overhead during decoding, the
+  /// benefits of saving space on the source pool data outweigh the overhead.
+  /// @param version The version of the source token data payload.
+  /// @param sourceTokenDataPayload The source token data payload to encode.
+  /// @return The encoded source token data payload.
+  function _encodeSourceTokenDataPayloadV0(
+    bytes4 version,
+    USDCTokenPool.SourceTokenDataPayloadV0 memory sourceTokenDataPayload
+  ) internal pure returns (bytes memory) {
+    /// Using abi.encodePacked() saves ~80 bytes on the source pool data by not using unnecessary padding.
+    /// abi.encode() = 96 bytes (32 + 32 + 32)
+    /// abi.encodePacked() = 16 bytes (4 + 8 + 4)
+    return abi.encodePacked(version, sourceTokenDataPayload.nonce, sourceTokenDataPayload.sourceDomain);
+  }
+
+  /// @notice Encodes the source token data payload into a bytes array.
+  /// @param version The version of the source token data payload.
+  /// @param sourceTokenDataPayload The source token data payload to encode.
+  /// @return The encoded source token data payload.
+  function _encodeSourceTokenDataPayloadV1(
+    bytes4 version,
+    USDCTokenPoolCCTPV2.SourceTokenDataPayloadV1 memory sourceTokenDataPayload
+  ) internal pure returns (bytes memory) {
+    /// Using abi.encodePacked() saves ~56 bytes on the source pool data by not using unnecessary padding.
+    /// abi.encode() = 96 bytes (32 + 32 + 32)
+    /// abi.encodePacked() = 40 bytes (4 + 4 + 32)
+    return abi.encodePacked(version, sourceTokenDataPayload.sourceDomain, sourceTokenDataPayload.depositHash);
+  }
+
+  /// @notice Decodes the abi.encodePacked() source pool data into its corresponding SourceTokenDataPayload struct.
+  /// @param sourcePoolData The source pool data to decode in raw bytes.
+  /// @return sourceTokenDataPayload The decoded source token data payload.
+  function _decodeSourceTokenDataPayloadV1(
+    bytes memory sourcePoolData
+  ) internal pure returns (USDCTokenPoolCCTPV2.SourceTokenDataPayloadV1 memory sourceTokenDataPayload) {
+    uint256 offset = 0;
+
+    // Since memory arrays cannot be sliced in the same way as calldata arrays, we need to create new bytes arrays
+    // to store the individual fields and then parse into their corresponding types.
```

Comment:
Individually iterating and copy-pasting byte by byte is super inefficient. Please use assembly to slice